scipy/stats/morestats.py

```diff
@@ -2167,24 +2229,37 @@ def ansari(x, y):
     uxy = unique(xy)
     repeats = (len(uxy) != len(xy))
     exact = ((m < 55) and (n < 55) and not repeats)
-    if repeats and (m < 55 or n < 55):
+    if mode == 'exact' and repeats:
+        raise ValueError("can't compute exact statistic with ties. "
+                         "Please use `mode='auto'` or `mode='approx'`.")
+    if mode == 'auto' and repeats and (m < 55 or n < 55):
         warnings.warn("Ties preclude use of exact statistic.")
-    if exact:
+    if (mode == 'exact' and not repeats) or (mode == 'auto' and exact):
         astart, a1, ifault = statlib.gscale(n, m)
         ind = AB - astart
         total = np.sum(a1, axis=0)
         if ind < len(a1)/2.0:
             cind = int(ceil(ind))
             if ind == cind:
-                pval = 2.0 * np.sum(a1[:cind+1], axis=0) / total
-            else:
+                cind = cind + 1
```

Comment:
Under what circumstances is `ind` not an integer?
The answer may have implications for the logic below. Just to be sure we agree - I think we want to make sure that the p-values always include the probability mass corresponding with the observed value of the test statistic, right? (If that doesn't make sense, let me know and I'll try to rephrase.)

Reply:
> Under what circumstances is `ind` not an integer?

Oh wow, I didn't think of this before. Thank you very much for noticing this! I checked and I think it is always an integer as `AB` will always be an integer (for `mode='exact'`) and `astart` is calculated as `ASTART = FPOINT((TEST + 1) / 2) * FPOINT(1 + TEST / 2)` where `TEST` is an integer and so all the divisions will be floor divisions. If this is the case, we can get rid of `cind` and `find` and just convert the `ind` itself to integer `ind = int(AB - astart)`. I tested that and it works. So, I will push that for review.

> Just to be sure we agree - I think we want to make sure that the p-values always include the probability mass corresponding with the observed value of the test statistic, right? (If that doesn't make sense, let me know and I'll try to rephrase.)

It makes sense and you are right.

Reply:
Oops; I didn't mean to ask for a change to that. Presumably the original authors had a reason for what they had? How confident are you? Changing existing code like that changes the scope of review a bit.

What I was going to say is that if `cind` is the index corresponding with the observed value of the test statistic, I thought we should leave it alone rather than redefine it `cind = cind + 1`. You had something like

```python3
cind = cind + 1
...
pval = 2.0 * np.sum(a1[:cind], axis=0) / total
```

It makes more sense to me to only add 1 when appropriate to the calculation. 
```python3
...
pval = 2.0 * np.sum(a1[:cind+1], axis=0) / total
```

Which is what you have now (good).

Reply:
> Oops; I didn't mean to ask for a change to that. Presumably the original authors had a reason for what they had? How confident are you? Changing existing code like that changes the scope of review a bit.

Oh, Sorry! I am not sure why the original authors did that. There may be a subtle case I may have missed. In that case, I will revert to the original behavior while adding one to `cind` only when required by the calculation. Does that sound like a correct thing to do here?